### DISCUSSION TITLE
How can I split `build.sc` into multiple files?

### DISCUSSION THREAD

**Discussion by micheal-hill** [com-lihaoyi/mill#1293] [Q&A] [answered]
I'm not sure if this is a question that's already answered elsewhere, but I didn't come up with much after searching through docs and the discussion forum. I'm trying to convert some workflow from `sbt` to `mill`

Is it possible to split my `build.sc` into multiple files? I'd like to have a hierarchy something like:
```
root
│  build.sc
└─ project             // arbitrary name; but indicates code that is important for the build script/config
│    utils.sc          // [1] contains some helper functions that are used by both `CommonModules.sc` and `build.sc`
│    CommonModules.sc  // [2] contains `trait ScalaCommon extends ScalaModule`
└─ app-module          // arbitrary name; represents a submodule of the project
     build.sc          // [3] module definition that is specific to this submodule
```

So far, it seems that:
1. Can be done using ammonite's import syntax `import $file.project.utils`
2. The import itself works, however it requires a second import to obtain `ScalaCommon` from the file. More importantly, it seems to break mill if I try to use it; resulting in an error: `Modules, Targets and Commands can only be defined within a mill Module`
3. I couldn't figure out a way to make this work at all

In each of the above situations, I've tried both `$cp` and `$file` syntax imports from ammonite - not sure if there's something else I should try too.

**Comment by lefou** [comment, marked as answer]
>     2. The import itself works, however it requires a second import to obtain `ScalaCommon` from the file. More importantly, it seems to break mill if I try to use it; resulting in an error: `Modules, Targets and Commands can only be defined within a mill Module`

The second import is normal, but the error not. Defining trait which contains targets should work. Only objects (the real modules in mill) need to be located in the `build.sc`.

To get to the cause of your specific error, you might need to share some code.

>     3. I couldn't figure out a way to make this work at all

Mill supports [foreign modules](https://com-lihaoyi.github.io/mill/mill/Modules.html#_foreign_modules) which should the way to support your project layout. Note, that you may not be able to refer back to the importing `build.sc` from the foreign module (in terms of tasks/targets). Also, I have to say I've never tried this myself and @lihaoyi once said, that foreign modules are much less used and therefore less tested than all the other stuff.

**Reply by itegebo** [reply]
When splitting, do all files need to have the `.sc` suffix?  Also, what's the reason for a different suffix?

**Reply by lefou** [reply]
Yes, all files need the `.sc` suffix. `.sc` is typically used for Scala scripts, e.g. in Ammonite but also for workbook files (Eclipse (scala-ide), IntelliJ IDEA), whereas `.scala` is Scala source code to be fed to the compiler.
